### PR TITLE
support basic pagination

### PR DESCRIPTION
**Pull request by noamhammer**
WIP

### CODE REVIEW

**Review by yanaise — APPROVED** [review]
LGTM

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 1]
```diff
@@ -115,7 +116,12 @@ const buildMultiEnvSource = (
     const allActiveElements = awu(_.values(getActiveSources(env)))
       .flatMap(async s => (s ? s.getAll() : awu([])))
     const { errors, merged } = await mergeElements(allActiveElements)
-    const elements = new InMemoryRemoteElementSource(createRemoteMap(getRemoteMapNameSpace('merged')))
+    const elements = new InMemoryRemoteElementSource(await createRemoteMap({
+      namespace: getRemoteMapNamespace('merged'),
+      serialize: (element: Element) => serialize([element]),
+      // TODO: we might need to pass static file reviver to the deserialization func
+      deserialize: async (data: string) => (await deserialize(data))[0],
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:123]: There's an issue hiding here with this [0]. The current `deserialize` implementation was done for multiple elements `deserialization` and we are using it for something that must be a single one.
In this case if someone did a mistake and saved multiple elements data then the elements that are not the first one will be lost silently (we didn't check there's only one here).

We might want to either check that `deserialize(data)` returned only one element or alternatively create a `deserializeElement` func that always returns exactly one Element and validates itself.

File: packages/workspace/src/merger/internal/common.ts [hunk 2]
```diff
@@ -37,10 +32,14 @@ export abstract class MergeError extends types.Bean<Readonly<{
   toString(): string {
     return this.message
   }
+
```
Comment (adamro) [inline packages/workspace/src/merger/internal/common.ts:35]: Maybe it's better to implement a default deserialize here and have whoever needs it to override it, now it looks like it's not a must while it basically is just as much as serialize is, you just went around it with extra code and the `MERGE_ERROR_NAME_TO_DESERIALIZE_FUNC `

File: packages/workspace/src/merger/internal/primitives.ts [hunk 3]
```diff
@@ -32,6 +33,19 @@ export class MultiplePrimitiveTypesUnsupportedError extends MergeError {
...
+  serialize = (): string => JSON.stringify({
+    type: MultiplePrimitiveTypesUnsupportedError.name,
+    args: { elemID: this.elemID.getFullName(), duplicates: serialize(this.duplicates) },
+  })
+
+  static deserialize = async (data: Values): Promise<MultiplePrimitiveTypesUnsupportedError> => {
+    const { elemID, duplicates } = data
+    return new MultiplePrimitiveTypesUnsupportedError({
+      elemID,
+      duplicates: (await deserialize(duplicates)) as PrimitiveType[],
```
Comment (adamro) [inline packages/workspace/src/merger/internal/primitives.ts:46]: Not sure it's critical but maybe assert/validate it's actually a PrimitiveType instead of casting?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 4]
```diff
@@ -214,17 +222,40 @@ export const getParsedNaclFiles = async (
 
 type buildNaclFilesStateResult = { state: NaclFilesState; changes: Change<Element>[] }
 
-const buildNaclFilesState = async (
-  newNaclFiles: ParsedNaclFile[],
-  createRemoteMap: RemoteMapCreator<Value>,
+const buildNaclFilesState = async ({
+  newNaclFiles, createRemoteMap, existingState, staticFileSource, sourceName,
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:226]: Why's the `new` needed in the name? `pasrsedNaclFiles` maybe?

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 5]
```diff
@@ -214,17 +222,40 @@ export const getParsedNaclFiles = async (
...
+}: {
+  newNaclFiles: ParsedNaclFile[]
+  createRemoteMap: RemoteMapCreator<Value>
   existingState?: NaclFilesState
-): Promise<buildNaclFilesStateResult> => {
+  staticFileSource: StaticFilesSource
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:231]: nit: `staticFilesSource`

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 6]
```diff
@@ -214,17 +222,40 @@ export const getParsedNaclFiles = async (
...
 type buildNaclFilesStateResult = { state: NaclFilesState; changes: Change<Element>[] }
 
-const buildNaclFilesState = async (
-  newNaclFiles: ParsedNaclFile[],
-  createRemoteMap: RemoteMapCreator<Value>,
+const buildNaclFilesState = async ({
+  newNaclFiles, createRemoteMap, existingState, staticFileSource, sourceName,
+}: {
+  newNaclFiles: ParsedNaclFile[]
+  createRemoteMap: RemoteMapCreator<Value>
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:229]: nit: `remoteMapCreator`

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 7]
```diff
@@ -214,17 +222,40 @@ export const getParsedNaclFiles = async (
...
+      namespace: getRemoteMapNamespace('elements_index', sourceName),
+      serialize: (val: string[]) => JSON.stringify(val),
+      deserialize: data => JSON.parse(data),
+    }) as RemoteMap<string[]>,
+    mergeErrors: await createRemoteMap({
+      namespace: getRemoteMapNamespace('errors', sourceName),
+      serialize: (val: MergeError[]) => serializeMergeErrors(val),
+      deserialize: async data => deserializeMergeErrors(data),
+    }) as RemoteMap<MergeError[]>,
+    mergedElements: new InMemoryRemoteElementSource(await createRemoteMap({
```
Comment (adamro) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:245]: Should this be an `InMemoryRemoteElementSource`?

Comment (noamhammer) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:245]: The `InMemoryRemoteElementSource` should be renamed to `RemoteElementSource`

File: packages/workspace/src/merger/internal/common.ts (hunk 2 above)
Comment (noamhammer) [reply packages/workspace/src/merger/internal/common.ts:35]: Still not sure that it will make the code look better. @roironn Do you maybe have an idea how to make this code looks better? (pretty sure that there is a typescript trick that I am missing)

File: packages/workspace/src/workspace/remote_map/remote_map.ts [hunk 8]
```diff
@@ -1,31 +0,0 @@
-/*
```
Comment (roironn) [inline packages/workspace/src/workspace/remote_map/remote_map.ts:1]: I may be missing something fairly obvious - but why was this deleted?

File: packages/core/src/local-workspace/remote_map.ts [hunk 9]
```diff
@@ -79,128 +85,196 @@ async function *aggregatedIterable(iterators: rocksdb.Iterator[]): AsyncIterable
...
+      if (!(loc in dbConnections)) {
+        db = rocksdb(loc)
+        dbConnections[loc] = db
+        await promisify(db.open.bind(db))()
+        await clearImpl(TEMP_PREFIX)
+      } else {
+        db = dbConnections[loc]
+      }
+    }
+    await createDBIfNotCreated(location)
```
Comment (roironn) [inline packages/core/src/local-workspace/remote_map.ts:203]: It wasn't change in this PR - but I'm still writing this:
I think we should move the `createDBIfNotCreate` method to be called in a lazy way whenever something it attempting to access the `db` var from the actual exposed methods:
1. It causes the `createRemoteMapCreator` function to be async - which will could cause difficulties if we'll try to create a remote map from inside a constructor.
2. Its a better design pattern for cases where the connection can get disconnected (since you can implement the `createDBIfNotCreated` so it recycle the connection. Its not something to be concerned for *right now* - but it a low effort change, and a better design DYT?

File: packages/core/src/local-workspace/remote_map.ts [hunk 10]
```diff
@@ -79,128 +85,196 @@ async function *aggregatedIterable(iterators: rocksdb.Iterator[]): AsyncIterable
...
-    let batch = db.batch()
-    for await (const element of elements) {
-      i += 1
-      cache.set(valueToKey(element), element)
-      batch.put(getAppropriateKey(valueToKey(element), temp), serialize(element))
-      if (i % mapOptions.batchInterval === 0) {
-        await promisify(batch.write.bind(batch))()
-        batch = db.batch()
-      }
+export const createRemoteMapCreator = <T>(location: string):
```
Comment (roironn) [inline packages/core/src/local-workspace/remote_map.ts:88]: Hmmm - not loving the "creator function for creator functions pattern".

Why do we need it? Why can't `location` be a part of the regular params passed to the remote map?

File: packages/workspace/src/merger/internal/common.ts (hunk 2 above)
Comment (roironn) [reply packages/workspace/src/merger/internal/common.ts:35]: I'm pretty sure using the default serializer would work as long as you'll add the errors classes to the `NameToType` variable in the serialiser. Not the *best* solution since it would require you to add each new error type to it - but its defiantly better then the current separate serialisation solution. 

(I actually like the concept of each class being responsible for its own ser/deser, but if we want to do it we need to go all in on it. Which we're not going to do right now)

File: packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts [hunk 11]
```diff
@@ -115,10 +124,15 @@ const buildMultiEnvSource = (
...
+    const elements = new RemoteElementSource(await remoteMapCreator({
+      namespace: getRemoteMapNamespace('merged'),
+      serialize: (element: Element) => serialize([element]),
+      // TODO: we might need to pass static file reviver to the deserialization func
+      deserialize: deserializeElement,
+    }))
     await elements.setAll(applyInstancesDefaults(
       merged.values(),
-      new InMemoryRemoteElementSource(merged)
+      new RemoteElementSource(merged)
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/multi_env/multi_env_source.ts:135]: Super nit: This is confusing! (Since we don't create a *remote* element source here, since merged is in memory. Maybe we just don't need the Remote in the name? Maybe just add a comment?)

File: packages/workspace/src/workspace/remote_map/remote_map.ts (hunk 8 above)
Comment (noamhammer) [reply packages/workspace/src/workspace/remote_map/remote_map.ts:1]: It's already declared in `workspace/remote_map.ts`

File: packages/core/src/local-workspace/remote_map.ts (hunk 10 above)
Comment (noamhammer) [reply packages/core/src/local-workspace/remote_map.ts:88]: Talked about it f2f in the zoom

File: packages/workspace/src/workspace/remote_map/remote_map.ts (hunk 8 above)
Comment (noamhammer) [reply packages/workspace/src/workspace/remote_map/remote_map.ts:1]: already declared at `workspace/remote_map.ts`
